### PR TITLE
Optimize SR Cnn algorithm

### PR DESCRIPTION
**Pull request by guinao**
We are excited to review your PR.

So we can do the best job, please check:

- [ ] There's a descriptive title that will make sense to other developers some time from now. 
- [ ] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [ ] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

Changes.
1. Adjust the expected value for not anomaly points.
2. Add minimum length checks for STL decompose to prevent exceptions. 

Associated Issues.
#5379

### CODE REVIEW

**Comment by guinao** [comment]
@mengaims is added to the review. #Closed

**Comment by guinao** [comment]
@harishsk is added to the review. #Closed

**Comment by guinao** [comment]
@lisahua is added to the review. #Closed

File: src/Microsoft.ML.TimeSeries/Deseasonality.cs [hunk 1]
```diff
@@ -120,12 +122,10 @@ public void Deseasonality(ref double[] values, int period, ref double[] results)
...
             }
+            // invoke the back up deseasonality method if stl decompose fails.
             else
             {
-                for (int i = 0; i < values.Length; ++i)
-                {
-                    results[i] = values[i];
-                }
+                _backupFunc.Deseasonality(ref values, period, ref results);
             }
```
Comment (harishsk) [inline src/Microsoft.ML.TimeSeries/Deseasonality.cs:129]: (comment not specific to this line)

I don't see any tests associated with this PR. What is the behavior you are expecting to change from this PR? Is it possible to add some test for that?

**Comment by guinao** [comment]
#5379

File: src/Microsoft.ML.TimeSeries/Deseasonality.cs (hunk 1 above)
Comment (guinao) [reply src/Microsoft.ML.TimeSeries/Deseasonality.cs:129]: The behavior change of this PR is to bound the value of not anomaly point inside the lower / upper boundary. Checks are added to test this behavior.

---
In reply to: [480434195](https://github.com/dotnet/machinelearning/pull/5374#discussion_r480434195) [](ancestors = 480434195)

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs [hunk 2]
```diff
@@ -709,10 +740,21 @@ public void AnomalyDetectionWithSrCnn(bool loadDataFromFile)
         {
             var ml = new MLContext(1);
             IDataView dataView;
+            var data = new List<TimeSeriesDataDouble>();
             var dataPath = GetDataPath("Timeseries", "period_anomaly.csv");
 
             // Load data from file into the dataView
-            dataView = ml.Data.LoadFromTextFile<TimeSeriesDataDouble>(dataPath, hasHeader: true);
+            StreamReader readFile = new StreamReader(dataPath);
+            string line;
```
Comment (harishsk) [inline test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs:748]: Any particular reason LoadFromTextFile doesn't work for you?

Comment (mstfbl) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs:748]: +1, `LoadFromTextFile` is meant for this exact purpose of loading from a data path with a class type provided. Is there a specific reason why you don't use `LoadFromTextFile`?

Comment (guinao) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs:748]: Basically, I want to enumerate the input data in the test to verify the input value is bounded by the output value 'upperBound' and 'lowerBound'. Could you point me an example that I can use the DataView to do this? Thanks.

---
In reply to: [482242678](https://github.com/dotnet/machinelearning/pull/5374#discussion_r482242678) [](ancestors = 482242678)

Comment (harishsk) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs:748]: You can use `CreateEnumerable`  to convert an IDataView into an enumerable. See [https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.dataoperationscatalog.createenumerable?view=ml-dotnet](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.dataoperationscatalog.createenumerable?view=ml-dotnet)

---
In reply to: [482675482](https://github.com/dotnet/machinelearning/pull/5374#discussion_r482675482) [](ancestors = 482675482,482242678)

Comment (guinao) [reply test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs]: Updated.

---
In reply to: [482718920](https://github.com/dotnet/machinelearning/pull/5374#discussion_r482718920) [](ancestors = 482718920,482675482,482242678)

**Review by harishsk — APPROVED** [review]
:shipit:

**Pull request merged**